funcs.php

```diff
@@ -0,0 +1,111 @@
+<?php
+
+function branches(
+    string $defaultBranch,
+    string $minimumCmsMajor,
+    // The following params are purely for unit testing, for the actual github action it will read json files instead
+    string $composerJson = '',
+    string $branchesJson = '',
+    string $tagsJson = ''
+) {
+    if (!is_numeric($defaultBranch)) {
+        throw new Exception('Default branch must be a number');
+    }
+    if (!ctype_digit($minimumCmsMajor)) {
+        throw new Exception('Minimum CMS major must be an integer');
+    }
+
+    // work out default major
+    preg_match('#^([0-9]+)+\.?[0-9]*$#', $defaultBranch, $matches);
+    $defaultMajor = $matches[1];
+    
+    // read __composer.json of the current (default) branch
+    $contents = $composerJson ?: file_get_contents('__composer.json');
+    $json = json_decode($contents);
+    if (is_null($json)) {
+        throw new Exception('Could not parse __composer.json');
+    }
+    $defaultCmsMajor = '';
+    $version = preg_replace('#[^0-9\.]#', '', $json->require->{'silverstripe/framework'} ?? '');
+    if (preg_match('#^([0-9]+)+\.?[0-9]*$#', $version, $matches)) {
+        $defaultCmsMajor = $matches[1];
+    } else {
+        $phpVersion = $json->require->{'php'} ?? '';
+        if (substr($phpVersion,0, 4) === '^7.4') {
+            $defaultCmsMajor = 4;
+        } elseif (substr($phpVersion,0, 4) === '^8.1') {
+            $defaultCmsMajor = 5;
+        }
+    }
+    if ($defaultCmsMajor === '') {
+        throw new Exception('Could not work out what the default CMS major version this module uses');
+    }
+    // work out major diff e.g for silverstripe/admin for CMS 5 => 5 - 2 = 3
+    $majorDiff = $defaultCmsMajor - $defaultMajor;
+
+    $minorsWithStableTags = [];
+    $contents = $tagsJson ?: file_get_contents('__tags.json');
+    foreach (json_decode($contents) as $row) {
+        $tag = $row->name;
+        if (!preg_match('#^([0-9]+)\.([0-9]+)\.([0-9]+)$#', $tag, $matches)) {
+            continue;
+        }
+        $minor = $matches[1] . '.' . $matches[2];
+        $minorsWithStableTags[] = $minor;
```

Comment:
```suggestion
        $minorsWithStableTags[$minor] = $minor;
```
There are gonna be a bunch of tags for each minor - cleaner to only include each minor once.

Reply:
Updated